Bottom: a540c52a792312c36c9cacfbfb23e1cf5c477e0e
Top:    47a1ff121d78e6106841040006b76dc86ea7b5f4
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-07-25 23:23:28 +0100

hid/gtk: Add a shutdown function to the renderer backends

This will be necessary to stop any event handlers / timers which
should not fire as the GUI is being shut down.

This code isn't actually very "live" at the moment, as the place I've
hooked up its trigger (the destroy event of the main window), doesn't
actually appear to get called.

This seems to be because the core "Quit" action just kills the program
dead, rather than attempting to shut down the GUI.


---

diff --git a/src/hid/gtk/gtkhid-gdk.c b/src/hid/gtk/gtkhid-gdk.c
index 4fdb8d3..25d38fd 100644
--- a/src/hid/gtk/gtkhid-gdk.c
+++ b/src/hid/gtk/gtkhid-gdk.c
@@ -1073,6 +1073,13 @@ ghid_init_renderer (int *argc, char ***argv, GHidPort *port)
 }
 
 void
+ghid_shutdown_renderer (GHidPort *port)
+{
+  g_free (port->render_priv);
+  port->render_priv = NULL;
+}
+
+void
 ghid_init_drawing_widget (GtkWidget *widget, GHidPort *port)
 {
 }
diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index ed34ea1..2d7fec0 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -822,6 +822,13 @@ ghid_init_renderer (int *argc, char ***argv, GHidPort *port)
 }
 
 void
+ghid_shutdown_renderer (GHidPort *port)
+{
+  g_free (port->render_priv);
+  port->render_priv = NULL;
+}
+
+void
 ghid_init_drawing_widget (GtkWidget *widget, GHidPort *port)
 {
   render_priv *priv = port->render_priv;
diff --git a/src/hid/gtk/gui-top-window.c b/src/hid/gtk/gui-top-window.c
index b11d32c..e9c2669 100644
--- a/src/hid/gtk/gui-top-window.c
+++ b/src/hid/gtk/gui-top-window.c
@@ -1851,6 +1851,7 @@ delete_chart_cb (GtkWidget * widget, GdkEvent * event, GHidPort * port)
 static void
 destroy_chart_cb (GtkWidget * widget, GHidPort * port)
 {
+  ghid_shutdown_renderer (port);
   gtk_main_quit ();
 }
 
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index 114574f..b0bbe0a 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -490,6 +490,7 @@ void ghid_invalidate_all ();
 void ghid_notify_crosshair_change (bool changes_complete);
 void ghid_notify_mark_change (bool changes_complete);
 void ghid_init_renderer (int *, char ***, GHidPort *);
+void ghid_shutdown_renderer (GHidPort *);
 void ghid_init_drawing_widget (GtkWidget *widget, GHidPort *);
 void ghid_drawing_area_configure_hook (GHidPort *port);
 void ghid_screen_update (void);
